toolchain/sem_ir/typed_nodes.h

```diff
@@ -0,0 +1,482 @@
+
+  Parse::Node parse_node;
+  // This node is an implementation detail of `StructType`, and doesn't produce
+  // a value, so has no type, even though it declares a field with a type.
+  StringId name_id;
+  TypeId field_type_id;
+};
+
+struct StructValue {
+  static constexpr auto Kind = NodeKind::StructValue.Define("struct_value");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  NodeId src_id;
+  NodeBlockId elements_id;
+};
+
+struct Temporary {
+  static constexpr auto Kind = NodeKind::Temporary.Define("temporary");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  NodeId storage_id;
+  NodeId init_id;
+};
+
+struct TemporaryStorage {
+  static constexpr auto Kind =
+      NodeKind::TemporaryStorage.Define("temporary_storage");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+};
+
+struct TupleAccess {
+  static constexpr auto Kind = NodeKind::TupleAccess.Define("tuple_access");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  NodeId tuple_id;
+  MemberIndex index;
+};
+
+struct TupleIndex {
+  static constexpr auto Kind = NodeKind::TupleIndex.Define("tuple_index");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  NodeId tuple_id;
+  NodeId index_id;
+};
+
+struct TupleInit {
+  static constexpr auto Kind = NodeKind::TupleInit.Define("tuple_init");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  NodeId src_id;
+  NodeBlockId elements_id;
+};
+
+struct TupleLiteral {
+  static constexpr auto Kind = NodeKind::TupleLiteral.Define("tuple_literal");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  NodeBlockId elements_id;
+};
+
+struct TupleType {
+  static constexpr auto Kind = NodeKind::TupleType.Define("tuple_type");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  TypeBlockId elements_id;
+};
+
+struct TupleValue {
+  static constexpr auto Kind = NodeKind::TupleValue.Define("tuple_value");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  NodeId src_id;
+  NodeBlockId elements_id;
+};
+
+struct UnaryOperatorNot {
+  static constexpr auto Kind = NodeKind::UnaryOperatorNot.Define("not");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  NodeId operand_id;
+};
+
+struct ValueAsReference {
+  static constexpr auto Kind =
+      NodeKind::ValueAsReference.Define("value_as_reference");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  NodeId value_id;
+};
+
+struct VarStorage {
+  static constexpr auto Kind = NodeKind::VarStorage.Define("var");
+
+  Parse::Node parse_node;
+  TypeId type_id;
+  StringId name_id;
+};
+
+// HasParseNode<T> is true if T has a `Parse::Node parse_node` field.
+template <typename T, typename ParseNodeType = Parse::Node T::*>
+constexpr bool HasParseNode = false;
+template <typename T>
+constexpr bool HasParseNode<T, decltype(&T::parse_node)> = true;
+
+// HasTypeId<T> is true if T has a `TypeId type_id` field.
+template <typename T, typename TypeIdType = TypeId T::*>
+constexpr bool HasTypeId = false;
+template <typename T>
+constexpr bool HasTypeId<T, decltype(&T::type_id)> = true;
```

Comment:
It might be worth moving these to the top, since otherwise they might be easily missed beneath all the actual typed nodes. I can understand though if you'd rather not; it only really makes sense if we don't have too much extra code like this.

Reply:
These to me feel like implementation details that most readers of this file won't be interested in, so I think I'd prefer to keep them here, or maybe move them to a different file?

Reply:
FYI, I agree this feels secondary to the typed nodes themselves. I'm not sure if I would consider it an implementation detail, because they are named by consumers of the header, but I would be fine with them at the end or in a different file. If it was in a different file, that file would naturally also contain `TypedNodeArgsInfo`.